system/lib/libc/dynlink.c

```diff
@@ -484,41 +484,55 @@ static void dlopen_onerror(struct dso* dso, void* user_data) {
 }
 
 // Modified version of path_open from musl/ldso/dynlink.c
-static int path_find(const char *name, const char *s, char *buf, size_t buf_size) {
+static int path_find(const char *name, int ncandidates, const char **candidates, char *buf, size_t buf_size) {
   size_t l;
   int fd;
-  for (;;) {
-    s += strspn(s, ":\n");
-    l = strcspn(s, ":\n");
-    if (l-1 >= INT_MAX) return -1;
-    if (snprintf(buf, buf_size, "%.*s/%s", (int)l, s, name) < buf_size) {
-      dbg("dlopen: path_find: %s", buf);
-      struct stat statbuf;
-      if (stat(buf, &statbuf) == 0 && S_ISREG(statbuf.st_mode)) {
-        return 0;
-      }
-      switch (errno) {
-      case ENOENT:
-      case ENOTDIR:
-      case EACCES:
-      case ENAMETOOLONG:
-        break;
-      default:
-        dbg("dlopen: path_find failed: %s", strerror(errno));
-        /* Any negative value but -1 will inhibit
-         * futher path search. */
-        return -2;
+  for (int c = 0; c < ncandidates; c ++) {
+    const char* s = candidates[c];
+    if (s == NULL) {
+      dbg("Candidate %d is null, skipping\n", c);
+      continue;
+    }
+    dbg("Candidate %d...\n", c);
+    for (;;) {
+      s += strspn(s, ":\n");
+      l = strcspn(s, ":\n");
+      if (l-1 >= INT_MAX) return -1;
+      if (snprintf(buf, buf_size, "%.*s/%s", (int)l, s, name) < buf_size) {
+        dbg("dlopen: path_find: %s", buf);
+        struct stat statbuf;
+        dbg("Checking file: %s\n", buf);
+        if (stat(buf, &statbuf) == 0 && S_ISREG(statbuf.st_mode)) {
+          dbg(" .. found\n");
+          return 0;
+        }
+        dbg(" .. not found\n");
+        switch (errno) {
+        case ENOENT:
+        case ENOTDIR:
+        case EACCES:
+        case ENAMETOOLONG:
+          break;
+        default:
+          dbg("dlopen: path_find failed: %s", strerror(errno));
+          /* Any negative value but -1 will inhibit
+          * futher path search. */
+          return -2;
+        }
       }
+      s += l;
     }
-    s += l;
   }
+  return -1;
 }
 
 // Resolve filename using LD_LIBRARY_PATH
-static const char* resolve_path(char* buf, const char* file, size_t buflen) {
+const char* _emscripten_resolve_path(char* buf, const char* rpath, const char* file, size_t buflen) {
```

Comment:
Can we name this something dylink specific?  How about `_emscripten_find_dylib` or `_emscripten_resolve_dylib_path`?